File: firebase/functions/lib/admin/payments/create_subscription_checkout_session.dart
import 'package:enum_to_string/enum_to_string.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'stripe_util.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/admin/billing_subscription.dart';
import 'package:data_models/community/community.dart';

/// Create a checkout session for the purchase of a given subscription type for a given community.
/// Returns a session id which should be handed off to Stripe client library which will handle the
/// redirects.
class CreateSubscriptionCheckoutSession
    extends OnCallMethod<CreateSubscriptionCheckoutSessionRequest> {
  CreateSubscriptionCheckoutSession()
      : super(
          'CreateSubscriptionCheckoutSession',
          (jsonMap) =>
              CreateSubscriptionCheckoutSessionRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    CreateSubscriptionCheckoutSessionRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final domain = functions.config.get('app.domain') as String;

    // For now, only allow community creator to create a new subscription
    final communitySnapshot = await firestore
        .document('community/${request.appliedCommunityId}')
        .get();
    orElseUnauthorized(
      communitySnapshot.exists,
      logMessage: 'Community to which subscription would apply not found',
    );
    final communityCreatorId =
        Community.fromJson(communitySnapshot.data.toMap()).creatorId;
    orElseUnauthorized(
      communityCreatorId == context.authUid,
      logMessage: 'User is not billing manager',
    );

    // Don't allow multiple subscriptions for the same user and community; should modify existing
    final subscriptionSnapshot = await firestore
        .collection('/stripeUserData/${context.authUid}/subscriptions')
        .where(BillingSubscription.kFieldCanceled, isEqualTo: false)
        .where(
          BillingSubscription.kFieldAppliedCommunityId,
          isEqualTo: request.appliedCommunityId,
        )
        .get();
    orElseUnauthorized(
      subscriptionSnapshot.isEmpty,
      logMessage:
          'Resumable subscription already found for this community and billing manager',
    );

    final customerId =
        await stripeUtil.getOrCreateCustomerStripeId(uid: context.authUid!);
    final priceId =
        await _getProductPrice(EnumToString.convertToString(request.type));

    final Map<String, String> params = {
      'success_url': 'https://$domain/${request.returnRedirectPath}',
      'cancel_url': 'https://$domain/${request.returnRedirectPath}',
      'payment_method_types[0]': 'card',
      'line_items[0][price]': priceId,
      'line_items[0][quantity]': '1',
      'mode': 'subscription',
      'customer': customerId,
      'subscription_data[metadata][appliedCommunityId]':
          request.appliedCommunityId,
    };

    final jsonResponse =
        await stripeUtil.post(path: '/checkout/sessions', params: params);

    final String sessionId = jsonResponse['id'];
    return CreateSubscriptionCheckoutSessionResponse(sessionId: sessionId)
        .toJson();
  }

  /// Given a plan type, return the id of the Stripe price object associated with it
  Future<String> _getProductPrice(String type) async {
    final productsResponse = await stripeUtil.get(path: '/products');
    final products = productsResponse["data"] as List;
    final product = products
        .where((product) => product["metadata"]["plan_type"] == type)
        .firstOrNull;
    orElseNotFound(
      product != null,
      logMessage: 'Stripe product not found for plan type: $type',
    );
    final productId = product['id'];

    final pricesResponse =
        await stripeUtil.get(path: '/prices?active=true&product=$productId');
    final prices = pricesResponse["data"] as List;
    orElseNotFound(
      prices.isNotEmpty,
      logMessage: 'Stripe price not found for product: $productId',
    );
    return prices[0]['id'];
  }

  // Sets minimum instances to 0
  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 0,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/admin/payments/create_subscription_checkout_session.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { stripeUtil } from './stripe_util';
import { orElseUnauthorized, orElseNotFound } from '../../utils/utils';

interface CreateSubscriptionCheckoutSessionRequest {
  appliedCommunityId: string;
  returnRedirectPath: string;
  type: string;
}

export class CreateSubscriptionCheckoutSession extends OnCallMethod<CreateSubscriptionCheckoutSessionRequest> {
  constructor() {
    super(
      'CreateSubscriptionCheckoutSession',
      (jsonMap) => jsonMap as CreateSubscriptionCheckoutSessionRequest,
      { runWithOptions: { timeoutSeconds: 60, memory: '1GB', minInstances: 0 } }
    );
  }

  async action(
    request: CreateSubscriptionCheckoutSessionRequest,
    context: functions.https.CallableContext
  ): Promise<Record<string, unknown>> {
    orElseUnauthorized(context.auth?.uid != null);

    const domain = functions.config().app?.domain as string;

    const communitySnap = await firestore.doc(`community/${request.appliedCommunityId}`).get();
    orElseUnauthorized(communitySnap.exists, { logMessage: 'Community to which subscription would apply not found' });
    const community = communitySnap.data() as Record<string, unknown>;
    orElseUnauthorized(
      community['creatorId'] === context.auth?.uid,
      { logMessage: 'User is not billing manager' }
    );

    const subscriptionSnap = await firestore
      .collection(`/stripeUserData/${context.auth!.uid}/subscriptions`)
      .where('canceled', '==', false)
      .where('appliedCommunityId', '==', request.appliedCommunityId)
      .get();
    orElseUnauthorized(subscriptionSnap.empty, { logMessage: 'Resumable subscription already found for this community and billing manager' });

    const customerId = await stripeUtil.getOrCreateCustomerStripeId({ uid: context.auth!.uid });
    const priceId = await this._getProductPrice(request.type);

    const params: Record<string, string> = {
      'success_url': `https://${domain}/${request.returnRedirectPath}`,
      'cancel_url': `https://${domain}/${request.returnRedirectPath}`,
      'payment_method_types[0]': 'card',
      'line_items[0][price]': priceId,
      'line_items[0][quantity]': '1',
      'mode': 'subscription',
      'customer': customerId,
      'subscription_data[metadata][appliedCommunityId]': request.appliedCommunityId,
    };

    const jsonResponse = await stripeUtil.post({ path: '/checkout/sessions', params });
    return { sessionId: jsonResponse['id'] };
  }

  private async _getProductPrice(type: string): Promise<string> {
    const productsResponse = await stripeUtil.get({ path: '/products' });
    const products = (productsResponse['data'] as unknown[]) ?? [];
    const product = products.find((p: any) => p['metadata']?.['plan_type'] === type);
    orElseNotFound(product != null, { logMessage: `Stripe product not found for plan type: ${type}` });
    const productId = (product as any)['id'] as string;

    const pricesResponse = await stripeUtil.get({ path: `/prices?active=true&product=${productId}` });
    const prices = ((pricesResponse['data'] as unknown[]) ?? []).sort(
      (a: any, b: any) => a['unit_amount'] - b['unit_amount']
    );
    orElseNotFound(prices.length > 0, { logMessage: `Pricing data not set in Stripe for product: ${productId}` });
    return (prices[0] as any)['id'] as string;
  }
}
